timemachine/cpp/src/kernels/k_flat_bottom_bond.cuh

```diff
@@ -1,6 +1,65 @@
 #include "../fixed_point.hpp"
 #include "k_fixed_point.cuh"
 
+// branchless implementation of piecewise function
+template <typename RealType>
+RealType __device__ __forceinline__ compute_flat_bottom_energy(
+    RealType k, RealType r_lt_rmin, RealType r_gt_rmax, RealType r, RealType rmin, RealType rmax) {
+    return (k / 4) * (r_lt_rmin * (pow(r - rmin, 4)) + r_gt_rmax * (pow(r - rmax, 4)));
+}
+
+template <typename RealType>
+void __global__ k_log_probability_selection(
+    const int N,                             // Num atoms
+    const double kBT,                        // BOLTZ * temperature
+    const float radius,                      // Radius, corresponds to r_max for flat bottom
+    const float k,                           // Constant restraint value
+    const unsigned int reference_idx,        // Idx that the probability is specific to
+    const double *__restrict__ coords,       // [N, 3]
+    const double *__restrict__ box,          // [3, 3]
+    const float *__restrict__ probabilities, // [N] probabilities of selection
+    unsigned int *__restrict__ selected      // [N] idx array, N if idx is not selected, else idx of coordinate
+) {
+    const unsigned int idx = blockDim.x * blockIdx.x + threadIdx.x;
+    if (idx >= N) {
+        return;
+    }
+    const float radius_sq = radius * radius;
+    const RealType bx = box[0 * 3 + 0];
+    const RealType by = box[1 * 3 + 1];
+    const RealType bz = box[2 * 3 + 2];
+
+    const RealType inv_bx = 1 / bx;
+    const RealType inv_by = 1 / by;
+    const RealType inv_bz = 1 / bz;
+
+    RealType atom_atom_dx = coords[idx * 3 + 0] - coords[reference_idx * 3 + 0];
+    RealType atom_atom_dy = coords[idx * 3 + 1] - coords[reference_idx * 3 + 1];
+    RealType atom_atom_dz = coords[idx * 3 + 2] - coords[reference_idx * 3 + 2];
+
+    atom_atom_dx -= bx * nearbyint(atom_atom_dx * inv_bx);
+    atom_atom_dy -= by * nearbyint(atom_atom_dy * inv_by);
+    atom_atom_dz -= bz * nearbyint(atom_atom_dz * inv_bz);
+
+    const RealType distance_sq =
+        atom_atom_dx * atom_atom_dx + atom_atom_dy * atom_atom_dy + atom_atom_dz * atom_atom_dz;
+
+    RealType prob = 1.0;
+    if (distance_sq >= radius_sq) {
+        RealType energy = compute_flat_bottom_energy<RealType>(
```

Comment:
Curiosity:
(1) should `k_log_probability_selection` be moved somewhere local md specific rather than flat-bottom bond specific,
(2) would it be hard to swap in other restraint functions besides this specific "flat bottom bond" in the future?

(The x^4 was a specific way to preempt a concern about the more widely used x^2 flat-bottom restraint raised about a prototype implementation, but I'm not aware if the concern has also arisen in practice / not sure this specific way of resolving the concern is an intrinsic part of local MD. Initially I think the logic of `k_log_probability_selection` is independent of the implementation details of the restraint function, and should remain mostly unchanged for any choice of non-negative valued restraint function, as in https://github.com/proteneer/timemachine/pull/785 )

Reply:
1) Keeping `k_log_probability_selection` here was intended to be very explicit about its implementation being specific to the flat bottom restraint. Don't have a strong opinion on this, could be moved elsewhere.

2) Would be pretty easy to swap out other restraint functions, would mostly depend on the ability to construct the restraint on the fly in C++. 

 `k_log_probability_selection` could probably be templated to use different energy functions in the future. Are you thinking we should support HarmonicBond and FlatBottomBond as different restraints?

Reply:
(1) Fair enough!

(2) Agreed -- I think that's the main requirement.

>k_log_probability_selection could probably be templated to use different energy functions in the future. Are you thinking we should support HarmonicBond and FlatBottomBond as different restraints?

Fine as-is. I didn't mean to advocate a templated or programmable choice among more than one option -- and I don't think there is more than one suitable option currently available. HarmonicBond is probably not immediately suitable.

Just trying to think ahead to what would be involved in revising any details of the implementation that are somewhat arbitrary / tentative (such as the functional form of the restraint).